ServiceEntry deletion - trigger endpoint shard update when some endpoints are remaining

ivpr opened the pull request:
…ints are remaining

**Please provide a description of this PR:**

istio-testing commented:
Hi @ivpr. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

ivpr commented:
Fixes #43853

hzxuzhonghu commented:
/ok-to-test

istio-testing commented:
@ivpr: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 14033648f4e4809bc5428c6cc4cfbdce3c25cc42 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/release-notes_istio/1633798068358877184) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented on pilot/pkg/serviceregistry/serviceentry/controller.go:
```diff
@@ -402,12 +402,16 @@ func (s *Controller) serviceEntryHandler(_, curr config.Config, event model.Even
 			s.XdsUpdater.SvcUpdate(shard, string(svc.Hostname), svc.Attributes.Namespace, model.EventDelete)
+		} else {
+			// If there are some endpoints remaining for the host, trigger eds cache update
+			partiallyDeletedSvcs = append(partiallyDeletedSvcs, svc)
```
nit: s/partiallyDeletedSvcs/updatedSvcs/g

ymesika replied:
It's not a nit.. Appending to existing `updatedSvcs` will clean up a lot of new code here.

hzxuzhonghu reviewed (commented):
Would be better to add tests to cover SE with same hosts scenario

istio-testing commented:
@ivpr: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 74c4c159905a277535c8d5079f427cbcf501b3db | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/release-notes_istio/1633827257707401216) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented on pilot/pkg/serviceregistry/serviceentry/controller.go:
```diff
@@ -408,6 +408,9 @@ func (s *Controller) serviceEntryHandler(_, curr config.Config, event model.Even
 		// Delete endpoint shards only if there are no service instances.
 		if len(s.serviceInstances.getByKey(instanceKey)) == 0 {
 			s.XdsUpdater.SvcUpdate(shard, string(svc.Hostname), svc.Attributes.Namespace, model.EventDelete)
+		} else {
```
I think this wont work. UpdatedSvcs are already processed (SvcUpdate event is already called at L400)

ivpr replied:
@ramaraochavali SvcUpdate is currently only incrementing the metric counters when the event is not `model.EventDelete`. Though functionally it shouldn't affect, the metric wouldn't be incremented. Let me know if that's a concern

Also the main intention of adding this to updatedSvcs is to trigger eds cache update `doEdsCacheUpdate` for the host

hzxuzhonghu replied:
should work @ramaraochavali

ramaraochavali replied:
>  Though functionally it shouldn't affect, the metric wouldn't be incremented. Let me know if that's a concern

Yes. This is what I meant. It would be little confusing

ramaraochavali reviewed (changes requested):
Please fix it and add test case

istio-testing commented:
@ivpr: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 8b6c539975fa3f0ac8bb79923b59bbb65a3b2e44 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/release-notes_istio/1633845531698532352) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ivpr commented:
@hzxuzhonghu & @ramaraochavali I've added the test cases for this
The test case scenario was already present but we were not doing any assertions on edscache events. I've added the assertions for eds cache in all places and verified the tests to running
Below is the assertion failing without the code fix
https://github.com/istio/istio/pull/43852/files#diff-db93d0d3aef31b582e2969e9b40dcdce6b5f7d23172876ffec7f2c6285c81519R414

Please review

ramaraochavali commented:
Can you fix lint?

istio-testing commented:
@ivpr: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio | 8b6c539975fa3f0ac8bb79923b59bbb65a3b2e44 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/lint_istio/1633845530033393664) | true | `/test lint`
gencheck_istio | 8b6c539975fa3f0ac8bb79923b59bbb65a3b2e44 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/gencheck_istio/1633845530851282944) | true | `/test gencheck`
release-notes_istio | a344358c64e78bfbf79ffd91a99916ec15733359 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/release-notes_istio/1634223209312161792) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ivpr: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 6450ea871f1285e65cb3710508dfbf00486dd3f6 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/release-notes_istio/1634238988669685760) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ivpr commented:
@ramaraochavali Fixed the lint errors. Had to amend and force push the commits to fix the author email for CLA. Please check

ramaraochavali commented:
LGTM. Please add release notes also

istio-testing commented:
@ivpr: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-telemetry-mc_istio | e555a49540cd79b1c3fe23867743855ff76236fc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/43852/integ-telemetry-mc_istio/1635253219955314688) | true | `/test integ-telemetry-mc`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ivpr commented:
/test integ-telemetry-mc

howardjohn commented on pilot/pkg/serviceregistry/serviceentry/controller_test.go:
```diff
@@ -714,7 +729,9 @@ func TestServiceDiscoveryWorkloadUpdate(t *testing.T) {
 		expectServiceInstances(t, sd, dnsSelector, 0, instances)
-		expectEvents(t, events, Event{kind: "xds"})
+		expectEvents(t, events,
+			Event{kind: "edscache", host: "dns.selector.com", namespace: dnsSelector.Namespace},
```
Why are we updating edscache for a DNS service entry? edscache is for EDS which isn't used for DNS service entries?

Is this intended or a side effect?

hzxuzhonghu replied:
It  is intended for  serviceentry with selector

hzxuzhonghu replied:
This is not necessary, but i found we didnot distinguish DNS address since very long before. So we always  update eds for them as well

howardjohn commented on pilot/pkg/serviceregistry/serviceentry/controller_test.go:
```diff
@@ -596,6 +608,7 @@ func TestServiceDiscoveryWorkloadUpdate(t *testing.T) {
 		expectServiceInstances(t, sd, selector, 0, instances)
 		expectEvents(t, events,
 			Event{kind: "svcupdate", host: "selector.com", namespace: selector.Namespace},
+			Event{kind: "edscache", host: "selector.com", namespace: selector.Namespace},
```
Similar comment, we do we need an EDS update here if we are going from 0 -> 0 instances? Do we need it or is it just a side effect?
